code/android-core-library/src/main/java/com/adobe/marketing/mobile/utils/RemoteDownloader.kt

```diff
@@ -0,0 +1,116 @@
+ * This allows easier path for components using this class to add queuing logic for jobs internally,
+ * maintaining compatibility when this class adds queuing logic in the future, and facilitating
+ * cleaner tests in composing components.
+ */
+class RemoteDownloader(
+    private val networkService: Networking,
+    private val cacheFileService: CacheFileService
+) {
+
+    companion object {
+        private const val TAG = "RemoteDownloader"
+    }
+
+    /**
+     * Represents a component that provides metadata about file content.
+     */
+    interface MetadataProvider {
+        companion object MetadataKeys {
+            internal const val HTTP_HEADER_IF_MODIFIED_SINCE = "If-Modified-Since"
+            internal const val HTTP_HEADER_IF_RANGE = "If-Range"
+            internal const val HTTP_HEADER_RANGE = "Range"
+            internal const val HTTP_HEADER_LAST_MODIFIED = "Last-Modified"
+            internal const val ETAG = "ETag"
+        }
+
+        /**
+         * Retrieves metadata for the [file] provided for the purpose of
+         * conditionally fetching content from the remote url.
+         *
+         * @param file the [File] for which metadata is needed
+         * @return the metadata of the [file] if it is valid;
+         *         emptymap if metadata cannot be computed
+         */
+        fun getMetadata(file: File?): Map<String, String>
+    }
+
+    /**
+     * Allows easier mocking during tests.
+     */
+    private val downloadJobSupplier: (Networking, CacheFileService, url: String, downloadSubDirectory: String?, MetadataProvider) -> RemoteDownloadJob =
+        { networking, fileCacheService, url, downloadDirectory, metadataProvider ->
+            RemoteDownloadJob(
+                networking,
+                fileCacheService,
+                url,
+                downloadDirectory,
+                metadataProvider
+            )
+        }
+
+    /**
+     * Triggers content download from [url] to [downloadSubDirectory] by using the [metadataProvider]
+     * to facilitate any partial downloads.
+     * Note that currently, triggers to download are independent of successive calls.
+     * [download] when called with the same url without waiting for a result via [completionCallback]
+     * will cause ambiguous results.
+     *
+     * @param url the url from which the content must be downloaded from
+     * @param downloadSubDirectory optional directory within the default cache directory used by [cacheFileService]
+     *        into which the rules must be placed
+     * @param metadataProvider the metadata provider that will be used to fetch conditional headers when making a
+     *        download request
+     * @param completionCallback the callback to notify about downloaded result
+     */
+    fun download(
+        url: String,
+        downloadSubDirectory: String?,
+        metadataProvider: MetadataProvider,
+        completionCallback: (File?) -> Unit
```

Comment:
Can we provide another parameter for the callback function that indicates the result of downloading, for example: the network is down, URL is not valid ...., then the client can decide if they need to retry it. 
Here is an example from the Swift side:
https://github.com/adobe/aepsdk-core-ios/blob/1f037918856101e863a8e6b1bc44e04ca40f189f/AEPCore/Sources/rules/RulesDownloader.swift#L50